pkg/segment/results/mresults/metricresults.go

```diff
@@ -702,14 +703,83 @@ func CalculateInterval(timerangeSeconds uint32) (uint32, error) {
 
 func (r *MetricsResult) aggregateFromAllTimeseries(aggregation structs.Aggregation) error {
 
+	var err error
 	switch aggregation.AggregatorFunction {
 	case segutils.Count:
 		r.computeAggCount(aggregation)
-	// Todo: Add TopK and BottomK
+	case segutils.TopK:
+		err = r.computeExtremesKElements(aggregation.FuncConstant, -1.0, true)
+	case segutils.BottomK:
+		err = r.computeExtremesKElements(aggregation.FuncConstant, 1.0, false)
 	default:
 		return fmt.Errorf("aggregateFromAllTimeseries: Unsupported aggregation: %v", aggregation)
 	}
 
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// The larger the priority, the earlier it will be popped out. Since we use the value as the priority, for `topk`, the larger the value, the more we want it to remain in the priority queue. Therefore, its priority should be smaller.
+// For bottomk, it's the opposite
+func (r *MetricsResult) computeExtremesKElements(funcConstant float64, factor float64, isTopK bool) error {
+	capacity := int(funcConstant)
+
+	if capacity <= 0 {
+		return fmt.Errorf("computeExtremesKElements: k must larger than 0")
+	}
+
+	// Use a PriorityQueue to store the top k elements for each timestamp, then separate the (timestamp, val) key-value pairs for each time series and generate the result
+	timestampToHeap := make(map[uint32]*utils.PriorityQueue)
+	for grpID, runningDS := range r.DsResults {
+		for i := 0; i < runningDS.idx; i++ {
+			timestamp := runningDS.runningEntries[i].downsampledTime
+
+			pq, exists := timestampToHeap[timestamp]
+			if !exists {
+				newPQ := make(utils.PriorityQueue, 0)
+				heap.Init(&newPQ)
+				heap.Push(&newPQ, &utils.Item{
+					Value:    grpID,
+					Priority: (runningDS.runningEntries[i].runningVal * factor),
+				})
+				timestampToHeap[timestamp] = &newPQ
+			} else {
+				if len(*pq) >= capacity {
+					item := heap.Pop(pq).(*utils.Item)
+					if isTopK && (item.Priority*factor) > runningDS.runningEntries[i].runningVal {
+						heap.Push(pq, item)
+						continue
+					}
+					if !isTopK && (item.Priority) < runningDS.runningEntries[i].runningVal {
+						heap.Push(pq, item)
+						continue
+					}
+				}
+				heap.Push(pq, &utils.Item{
+					Value:    grpID,
+					Priority: (runningDS.runningEntries[i].runningVal * factor),
+				})
+			}
+		}
+	}
+
+	for timestamp, pq := range timestampToHeap {
+		for pq.Len() > 0 {
+			item := heap.Pop(pq).(*utils.Item)
+			_, exists := r.Results[item.Value]
+			if !exists {
+				r.Results[item.Value] = make(map[uint32]float64)
+			}
+			r.Results[item.Value][timestamp] = (item.Priority * factor) // Restore to original value
```

Comment:
I think you actually want
```
(item.Priority / factor) // Restore to original value
```
Right? In your usage you only use `factor = 1` or `factor = -1`, so `*` and `/` are equivalent, but for a more general `factor` you should use division